client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -108,7 +108,29 @@ public void readerOffline(String readerId, Position lastPosition) {
 
     @Override
     public CompletableFuture<Checkpoint> initiateCheckpoint(String checkpointName, ScheduledExecutorService backgroundExecutor) {
-        synchronizer.updateStateUnconditionally(new CreateCheckpoint(checkpointName));
+
+        String rejectMessage = "rejecting checkpoint request since pending checkpoint reaches max allowed limit";
+
+        boolean canPerformCheckpoint = synchronizer.updateState((state, updates) -> {
+            ReaderGroupConfig config = state.getConfig();
+            CheckpointState checkpointState = state.getCheckpointState();
+            int maxOutstandingCheckpointRequest = config.getMaxOutstandingCheckpointRequest();
+            int currentOutstandingCheckpointRequest = checkpointState.getOutstandingCheckpoints();
+            if (currentOutstandingCheckpointRequest >= maxOutstandingCheckpointRequest) {
+                log.warn("maxOutstandingCheckpointRequest: {}, currentOutstandingCheckpointRequest: {}, errorMessage: {} {}",
```

Comment:
I can't remember whether we talked about throwing vs. returning false and logging. I'm worried that returning false is somewhat an ambiguous behavior. If a call to initiate checkpoint returns false, then the application is not sure whether this is a permanent or transient behavior. For example, there could some error condition such as a data corruption that prevents a new checkpoint from starting a checkpoint. I also need to check whether  a call to create checkpoint can throw, in which case we are already throwing exceptions when something doesn't work.

I feel that throwing an exception like `MaxExceededException` or `MaxNumberOfCheckpointsExceededException`.

Reply:
We are logging the error as well as throwing `CheckpointFailedException` when maximum checkpoint request exceeds ( https://github.com/pravega/pravega/blob/d805db8d6621b68d6ed360f1f933e855cd3a2690/client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java#L131). Are you suggesting to throw a different checked exception like `MaxNumberOfCheckpointsExceededException` instead of throwing `CheckpointFailedException`?